### ISSUE TITLE
[HOLD for payment 2022-04-01] [$1000] Long press on image doesn't open the context menu - Reported by @rushatgabhane

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7462]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app
2. Go to any chat with attachment image
3. Hard press on the image 

## Expected Result:
Long press on image should open the context menu.

## Actual Result:
Long press on image doesn't open the context menu. (Long press on side of image does)

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- iOS
- Android

**Version Number:** 1.1.33-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/151732112-f7a61dea-ed8f-4588-b646-40afbf49da50.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @rushatgabhane
**Slack conversation:**  https://expensify.slack.com/archives/C01GTK53T8Q/p1642351724013000

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tjferriss (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by stephanieelliott** [comment]
Was able to reproduce this on iOS. Posted to Upwork: https://www.upwork.com/jobs/~0109b6b91df96f2613

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Current assignee @roryabraham is eligible for the Exported assigner, not assigning anyone new.

**Comment by parasharrajat** [comment]
@rushatgabhane I have a few questions. 
What should be shown in the contextMenu? Do you expect to show the main menu the same as the message menu? 
We don't have any specifc options for images for now.

**Comment by parasharrajat** [comment]
Ok. Thanks for confirming.

**Comment by stephanieelliott** [comment]
2xd price to $500

**Comment by aswin-s** [comment]
**Reason:**

Image thumbnail is wrapped with in a TouchableOpacity, which captures all touch events at thumbnail level.
https://github.com/Expensify/App/blob/ddae43334629449e1cd4a9296fa18b234d7ccd94/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L47-L64

Context menu is opened via `longPress` action defined at ReportActionItem level, which is a parent component to `ImageRenderer`

https://github.com/Expensify/App/blob/9d6e07a0d3bb298d7edba16824efb95afc94f280/src/pages/home/report/ReportActionItem.js#L152-L157

Since the inner touchable captures the tap event it never reaches the `PressableWithSecondaryInteraction` HOC which fires the `onSecondaryInteraction` callback for showing popup. 


**Proposal:**

Move AttachmentModal logic to ReportActionItem level and control the visibility of the modal from onPress callback of `PressableWithSecondaryInteraction`.

```jsx
            <PressableWithSecondaryInteraction
                ref={el => this.popoverAnchor = el}
                onPressIn={() => this.props.isSmallScreenWidth && canUseTouchScreen() && ControlSelection.block()}
                onPressOut={() => ControlSelection.unblock()}
                onPress={() => this.showAttachmentModal}
                onSecondaryInteraction={this.showPopover}
                preventDefaultContentMenu={!this.props.draftMessage}

            >
```


We could go further ahead and introduce an AttachmentModal route at the RootNavigation level and navigate to the modal on tapping image thumbnail. This way a nested Modal inside each `ReportActionItem` could be removed.

**Comment by stephanieelliott** [comment]
Hey @parasharrajat, any feedback on the proposal above?

**Comment by parasharrajat** [comment]
Sorry, I was sleepy when I reviewed it so couldn't leave a comment.  

@aswin-s 

> Move AttachmentModal logic to ReportActionItem level and control the visibility of the modal from onPress callback of 

I don't see how will you do that in your proposal. 

> We could go further ahead and introduce an AttachmentModal route at the RootNavigation level and navigate to the modal on tapping image thumbnail. This way a nested Modal inside each ReportActionItem could be removed.

It will create more challenges. Do we want to make the attachment URL accessible? I think no. As attachments are related to reports. It would need to be sub root of the report.  I don't think this is a good approach.  

---

There are other ways. You can take a look at `BaseAnchorForCommentsOnly` to see one. There are a few challenges with that approach. Like MiniMenu should be hidden when the menu is opened. How will you highlight the row when the menu is opened? Other such things.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by stephanieelliott** [comment]
I'm about to go OOO, so reapplied the label to assign this to another CM team member (thanks @kadiealexander !)

**Comment by ahmdshrif** [comment]
# Proposal 

react-native-reanimated have component solve this. 
https://docs.swmansion.com/react-native-gesture-handler/docs/next/gesture-handlers/api/longpress-gh

we just need to wrap our PressableWithSecondaryInteraction component with it. 

in src/components/PressableWithSecondaryInteraction/index.native.js

```ts
        <LongPressGestureHandler
            onHandlerStateChange={(e) => {
                if (e.nativeEvent.state !== State.ACTIVE) {
                    return;
                }
                e.preventDefault();
                HapticFeedback.trigger();
                props.onSecondaryInteraction(e);
            }}
            minDurationMs={800}
        >
           ...
        </LongPressGestureHandler>
  ```

  
  https://user-images.githubusercontent.com/21364901/155994265-f96ecd2d-a3ec-4f06-87c4-f9095d431a3e.mov

**Comment by parasharrajat** [comment]
Nice and easy solution @ahmdshrif. I like it. 
> Image thumbnail is wrapped in a TouchableOpacity, which captures all touch events at thumbnail level.

Is this able to receive the touches even in situations where we have nested Touch handlers?

**Comment by ahmdshrif** [comment]
Yes, for  example when press on image it get open .

**Comment by parasharrajat** [comment]
Ok, Thanks. 

I like @ahmdshrif 's proposal. It is utilizing the existing package to fix the issue in clean way.  

cc: @roryabraham 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by roryabraham** [comment]
> I'd expect the context menu to have all other options i.e. unread, copy and delete.

Something I just thought of – copy should unfortunately be removed from the context menu until https://github.com/Expensify/App/issues/4629 is done.

**Comment by roryabraham** [comment]
Nice, LGTM 👍

**Comment by MelvinBot** [comment]
📣 @ahmdshrif You have been assigned to this job by @roryabraham!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by parasharrajat** [comment]
> Something I just thought of – copy should unfortunately be removed from the context menu until https://github.com/Expensify/App/issues/4629 is done.
Yeah, forgot to add that point.  

@ahmdshrif This needs to be done as well https://github.com/Expensify/App/issues/7462#issuecomment-1056005409.

**Comment by ahmdshrif** [comment]
@parasharrajat Ok i will do it . 

@kadiealexander i apply to the job .

**Comment by kadiealexander** [comment]
Offer sent, thank @ahmdshrif!

**Comment by ahmdshrif** [comment]
@kadiealexander thanks, I working on it. but look like there issue on offer 
this is the Upwork message to me. 

>This offer is not available anymore

**Comment by kadiealexander** [comment]
@ahmdshrif could you please share an update on where things are at?

**Comment by ahmdshrif** [comment]
@kadiealexander it's in review

**Comment by roryabraham** [comment]
Sorry, the delay here is on me. Catching up now...

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.43-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/8018


If no regressions arise, payment will be issued on 2022-03-24. :confetti_ball:

**Comment by parasharrajat** [comment]
@kadiealexander Could you please hold C+ payment for me on this issue until I request/ping you. I would like to wait for a week. Please put this on weekly so that you don't get notifications. Thanks.

**Comment by ahmdshrif** [comment]
@kadiealexander but it will be good if my payment not hold 😄

Thanks

**Comment by kadiealexander** [comment]
@ahmdshrif of course! I'll issue your payment and @rushatgabhane's reporting bonus today.

@parasharrajat I'll keep yours on hold, let me know when you're ready.

**Comment by kadiealexander** [comment]
Paid:

- [x] @ahmdshrif - Done! Thanks for your work Ahmed.
- [ ] @rushatgabhane - Sent you a contract to issue you the reporting bonus, please accept.
- [ ] @parasharrajat on hold as discussed.

**Comment by ahmdshrif** [comment]
Thanks @kadiealexander

**Comment by kadiealexander** [comment]
@parasharrajat how are you tracking for this? Still want me to keep it on hold?

**Comment by parasharrajat** [comment]
Nope @kadiealexander. You are good to go.

**Comment by kadiealexander** [comment]
@parasharrajat just sent you a contract, please accept!

**Comment by parasharrajat** [comment]
@kadiealexander Done.

**Comment by parasharrajat** [comment]
Bump @kadiealexander.

**Comment by kadiealexander** [comment]
Apologies @parasharrajat, I was ooo! All paid now :)
